craft_application/models/constraints.py

```diff
@@ -193,3 +195,44 @@ def validate(value: str) -> str:
 Applications may use a different set of constraints if necessary, but
 ideally they will retain this same constraint.
 """
+
+
+def _parse_spdx_license(value: str) -> license_expression.LicenseExpression:
+    licensing = license_expression.get_spdx_licensing()
+    if (
+        lic := licensing.parse(  # pyright: ignore[reportUnknownMemberType]
+            value, validate=True
+        )
+    ) is not None:
+        return lic
+    raise ValueError
+
+
+def _validate_spdx_license(value: str) -> str:
+    """Ensure the provided licence is a valid SPDX licence."""
+    try:
+        lic = _parse_spdx_license(value)
+    except (license_expression.ExpressionError, ValueError):
+        raise ValueError(
+            f"License '{value}' not valid. It must be in SPDX format.",
+        ) from None
+    else:
+        return str(lic.simplify())
+
+
+SpdxLicenseStr = Annotated[
+    str,
+    pydantic.AfterValidator(_validate_spdx_license),
+    pydantic.Field(
+        title="Licence",
+        description="SPDX license string.",
+        examples=[
+            "GPL-3.0",
+            "MIT",
+            "LGPL-3.0-or-later",
+            "GPL-3.0+ and MIT",
+        ],
+    ),
+]
+
+ProprietaryLicenseStr = Literal["proprietary"]
```

Comment:
Let's make a `LicenseStr` that's a union of these.

Reply:
Added in: https://github.com/canonical/craft-application/pull/402/commits/00ea50e734d5017049b2690cc8ef4d03ad8925a3